text/0000-variadic-generics.md

```diff
@@ -0,0 +1,209 @@
+programmers to abstract over non-homogeneous collections, and they make it
+possible to implement functions which accept arguments of varying length and
+type.
+
+Rust has a demonstrable need for variadic generics.
+
+In Rust's own standard library, there are a number of traits which have
+been repeatedly implemented for tuples of varying size up to length 12 using
+macros. This approach has several downsides:
+- It presents arbitrary restrictions on tuples of size 13+.
+- It increases the size of the generated code, resulting in slow compile times.
+- It complicates documentation
+(see the list of trait implementations in
+[this documentation](https://doc.rust-lang.org/std/primitive.tuple.html)).
+
+These arbitrary tuple-length restrictions, manual tuple macros, and confusing
+documentation all combine to increase Rust's learning curve.
+
+Furthermore, community library authors are required to implement similar
+macro-based approaches in order to implement traits for tuples. In the `Diesel`
+crate, it was discovered that replacing macro-generated tuple implementations
+with a structurally-recursive implementation (such as the one proposed here)
+resulted in a 50% decrease in the amount of code generated and a 70% decrease
+in compile times ([link](https://github.com/diesel-rs/diesel/pull/747)). This
+demonstrates that Rust's lack of variadic generics is resulting in a subpar
+edit-compile-debug cycle for at least one prominent, high-quality crate.
+
+The solution proposed here would resolve the limitations above by making it
+possible to implement traits for tuples of arbitrary length. This change would
+make Rust libraries easier to understand and improve the edit-compile-debug
+cycle when using variadic code.
+
+
+# Detailed design
+[design]: #detailed-design
+
+## The `Tuple` Trait
+The following would be implemented by all tuple types:
+```rust
+trait Tuple {
+    type AsRefs<'a>: Tuple + 'a;
+    type AsMuts<'a>: Tuple + 'a;
+    fn elements_as_refs<'a>(&'a self) -> Self::AsRefs<'a>;
+    fn elements_as_mut<'a>(&'a mut self) -> Self::AsMuts<'a>;
+}
+```
+
+The types `AsRefs` and `AsMuts` are the corresponding tuples of references to
+each element in the original tuple. For example,
+`(A, B, C)::AsRefs = (&A, &B, &C)` and
+`(A, B, C)::AsMuts = (&mut A, &mut B, &mut C)`
+
+The `Tuple` trait should only be implemented for tuples and marked with the
+`#[fundamental]` attribute described in
+[the coherence RFC](https://github.com/rust-lang/rfcs/blob/master/text/1023-rebalancing-coherence.md).
+This would allow coherence and type-checking to be extended to assume that no
+implementations of `Tuple` will be added. This enables an increased level of
+negative reasoning making it easier to write blanket implementations of traits
+for tuples.
+
+## The `(Head, ...Tail)` Type Syntax
+This syntax would allow for a `Cons`-cell-like representation of tuple types.
+For example, `(A, ...(B, C))` would be equivalent to `(A, B, C)`. This allows
```

Comment:
To be more clear: both tail-only and generalized single-repeat require the same primitive: a *single* check somewhere in the compiler that the `T` in `...T` is now known to be a tuple (i.e. after substituting in type parameters, resolving associated types, etc.) *therefore* expand it in place.
And concatenating 3 lists of types is hardly any more difficult than concatenating 2 lists.
This is done *only once*, mind you (probably should be done when creating any tuple IMO).

On top of that, everywhere a tuple is introspected now has to take into account a partially unknown tuple, and *that* has the same cost for both approaches, except if you do it one way then later you have to redo in the other way, and there can even be subtle bugs introduced from this.